core/src/position.rs

```diff
@@ -59,29 +68,54 @@ impl Default for PosIdx {
 /// An immutable table storing the position of values.
 #[derive(Debug, Default, PartialEq, Eq, Clone)]
 pub struct PosTable {
-    positions: Vec<TermPos>,
+    positions: Vec<RawSpan>,
 }
 
 impl PosTable {
     pub fn new() -> Self {
         // We always populate the first entry with `TermPos::None`, so that we can safely use the
         // index `0` (`PosIdx::NONE`) to mean unintialized position, without having to special case
         // the undefined position.
-        Self {
-            positions: vec![TermPos::None],
-        }
+        Self { positions: vec![] }
     }
 
     /// Inserts a new position for value and returns its index.
     pub fn push(&mut self, pos: TermPos) -> PosIdx {
-        let next = self.positions.len();
-        self.positions.push(pos);
-        PosIdx(u32::try_from(next).expect("maximum number of positions reached for values"))
+        match pos {
+            TermPos::Original(span) | TermPos::Inherited(span) => {
+                let next = self.positions.len();
+                self.positions.push(span);
+                // The conversion can't overflow, because this method is
+                // the only one that increases the length and it panics at
+                // `PosIdx::MAX`
+                let idx = next as u32;
+                if idx > PosIdx::MAX {
+                    panic!("maximum number of positions reached");
+                }
+                if matches!(pos, TermPos::Inherited(_)) {
+                    PosIdx(idx).to_inherited()
+                } else {
+                    PosIdx(idx)
+                }
+            }
+            TermPos::None => PosIdx::NONE,
+        }
     }
 
     /// Returns the position at index `idx`.
     pub fn get(&self, idx: impl Into<PosIdx>) -> TermPos {
```

Comment:
Unrelated, but I think we can get rid of this. This is a left-over from when we had `InlinePosIdx` which was convertible to `PosIdx`.

```suggestion
    pub fn get(&self, idx: PosIdx) -> TermPos {
```